pkg/reconciler/ingress/status.go

```diff
@@ -338,14 +351,45 @@ func (m *StatusProber) listVirtualServicePodIPs(vs *v1alpha3.VirtualService) ([]
 				continue
 			}
 
+			// Resolve the current gateway server's port to a name defined in the gateway's service.
+			portName, err := findNameForPortNumber(service, int32(server.Port.Number))
+			if err != nil {
+				return nil, fmt.Errorf("failed to find port name: %v", err)
+			}
 			for _, eps := range endpoints {
 				for _, sub := range eps.Subsets {
+					portNumber, err := findPortNumberForName(sub, portName)
+					if err != nil {
+						return nil, fmt.Errorf("failed to find port number: %v", err)
+					}
+
+					portStr := strconv.Itoa(int(portNumber))
 					for _, addr := range sub.Addresses {
-						podIPs = append(podIPs, net.JoinHostPort(addr.IP, strconv.Itoa(server.Port.Number)))
+						podIPs = append(podIPs, net.JoinHostPort(addr.IP, portStr))
 					}
 				}
 			}
 		}
 	}
 	return podIPs, nil
 }
+
+// findNameForPortNumber finds the name for a given port as defined by a Service.
```

Comment:
Do we need to do this two steps process InputPort -> Name, Name -> Endpoint Port?
Isn't this always going to be the same thing as InputPort -> TargetPort in Service Spec?

Reply:
It is, but `TargetPort` is `IntOrString` so it can actually point to a Pod's port name. Which brings us full-circle to how it looked before (with a pod informer).

Reading it from the pod is not much different and needs a loop through all pods as well.

Reply:
Ah, I forgot about that.